apps/aecore/lib/aecore/chain/header.ex

```diff
@@ -71,4 +87,106 @@ defmodule Aecore.Chain.Header do
   def base58c_decode(bin) do
     {:error, "#{__MODULE__}: Wrong data: #{inspect(bin)}"}
   end
+
+  @spec encode_to_binary(t()) :: binary()
+  def encode_to_binary(%Header{} = header) do
+    <<
+      header.version::@header_version_size,
+      header.height::@header_height_size,
+      header.prev_hash::binary-size(@header_hash_size),
+      header.txs_hash::binary-size(@txs_hash_size),
+      header.root_hash::binary-size(@root_hash_size),
+      header.target::@header_target_size,
+      pow_to_binary(header.pow_evidence)::binary-size(@pow_size),
+      header.nonce::@header_nonce_size,
+      header.time::@header_time_size,
+      header.miner::binary-size(@pubkey_size)
+    >>
+  end
+
+  def encode_to_binary(_) do
+    {:error, "#{__MODULE__}: Illegal header structure serialization"}
+  end
+
+  @spec decode_from_binary(binary()) :: {:ok, Header.t()} | {:error, String.t()}
+  def decode_from_binary(<<
+        version::@header_version_size,
+        height::@header_height_size,
+        prev_hash::binary-size(@header_hash_size),
+        txs_hash::binary-size(@txs_hash_size),
+        root_hash::binary-size(@root_hash_size),
+        target::@header_target_size,
+        pow_evidence_bin::binary-size(@pow_size),
+        nonce::@header_nonce_size,
+        time::@header_time_size,
+        miner::binary-size(@pubkey_size)
+      >>) do
+    with {:ok, pow_evidence} <- binary_to_pow(pow_evidence_bin) do
```

Comment:
The `with` statement is to check multiple valid functions, here we can simply use `case`